cpp/test/distance/sparse_l2_nn.cu

```diff
@@ -0,0 +1,494 @@
+    int num_groups = params.num_groups;
+
+    naive<DataT, Sqrt>(min_ref.data(),
+                       x.data(),
+                       y.data(),
+                       adj.data(),
+                       group_idxs.data(),
+                       m,
+                       n,
+                       k,
+                       num_groups,
+                       (int*)workspace.data(),
+                       stream);
+  }
+
+  void runTest(cub::KeyValuePair<int, DataT>* out)
+  {
+    int m          = params.m;
+    int n          = params.n;
+    int k          = params.k;
+    int num_groups = params.num_groups;
+
+    MinAndDistanceReduceOp<int, DataT> redOp;
+    sparseL2NN<DataT, cub::KeyValuePair<int, DataT>, int>(
+      out,
+      x.data(),
+      y.data(),
+      xn.data(),
+      yn.data(),
+      adj.data(),
+      group_idxs.data(),
+      num_groups,
+      m,
+      n,
+      k,
+      (void*)workspace.data(),
+      redOp,
+      raft::distance::KVPMinReduce<int, DataT>(),
+      Sqrt,
+      true,
+      stream);
+    RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
+  }
+};
+
+template <typename T>
+struct CompareApproxAbsKVP {
+  typedef typename cub::KeyValuePair<int, T> KVP;
+  CompareApproxAbsKVP(T eps_) : eps(eps_) {}
+  bool operator()(const KVP& a, const KVP& b) const
+  {
+    T diff  = raft::abs(raft::abs(a.value) - raft::abs(b.value));
+    T m     = std::max(raft::abs(a.value), raft::abs(b.value));
+    T ratio = m >= eps ? diff / m : diff;
+    return (ratio <= eps);
+  }
+
+ private:
+  T eps;
+};
+
+template <typename T>
+struct CompareExactKVP {
+  typedef typename cub::KeyValuePair<int, T> KVP;
+  bool operator()(const KVP& a, const KVP& b) const
+  {
+    if (a.value != b.value) return false;
+    return true;
+  }
+};
+
+template <typename K, typename V, typename L>
+::testing::AssertionResult devArrMatch(const cub::KeyValuePair<K, V>* expected,
+                                       const cub::KeyValuePair<K, V>* actual,
+                                       size_t size,
+                                       L eq_compare,
+                                       cudaStream_t stream = 0)
+{
+  typedef typename cub::KeyValuePair<K, V> KVP;
+  std::shared_ptr<KVP> exp_h(new KVP[size]);
+  std::shared_ptr<KVP> act_h(new KVP[size]);
+  raft::update_host<KVP>(exp_h.get(), expected, size, stream);
+  raft::update_host<KVP>(act_h.get(), actual, size, stream);
+  RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
+  for (size_t i(0); i < size; ++i) {
+    auto exp = exp_h.get()[i];
+    auto act = act_h.get()[i];
+    if (!eq_compare(exp, act)) {
+      return ::testing::AssertionFailure()
+             << "actual=" << act.key << "," << act.value << " != expected=" << exp.key << ","
+             << exp.value << " @" << i;
+    }
+  }
+  return ::testing::AssertionSuccess();
+}
+
+const std::vector<Inputs<float>> inputsf = {
+  {0.001f, 32, 32, 32, 2, 1234ULL, AdjacencyPattern::all_true},
```

Comment:
Not a huge deal but I'm wondering if it might make sense to also include a skewed degree distribution that might follow small-world or power-law characteristics (and it could be a good opportunity to use the rmat generator).

Reply:
The benchmark is quite heavy as is. The current patterns have mainly been chosen to highlight performance characteristics:
- all_true: what happens when we can't skip any computations
- all_false: when we can skip all computation (how fast is skipping, not really interesting in the real world)
- checkerboard_64: what if each CUDA block can skip exactly half the computations (expected 2x speedup)
- checkerboard_4: what if some warps (not sure anymore actually..) can skip computations (expected speedup: minor)
- checkerboard: what if some threads in a warp can skip computations (expected speedup: none)

I agree that sampling from a distribution would be more useful to get a "real-world" feel for the end-to-end performance, but it does not directly help when comparing small changes.